plugins/destination/clickhouse/client/migrate.go

```diff
@@ -12,67 +14,140 @@ import (
 
 // Migrate relies on the CLI/client to lock before running migration.
 func (c *Client) Migrate(ctx context.Context, tables schema.Tables) error {
-	current, err := c.getTableDefinitions(ctx)
+	currentSchema, err := c.getTableDefinitions(ctx)
 	if err != nil {
 		return err
 	}
 
+	newSchema := queries.NormalizedTables(tables)
+	if c.spec.MigrateMode != specs.MigrateModeForced {
+		nonSafeMigratableTables, changes := c.nonAutoMigrableTables(newSchema, currentSchema)
+		if len(nonSafeMigratableTables) > 0 {
+			return fmt.Errorf("tables %s with changes %v require force migration. use 'migrate_mode: forced'", strings.Join(nonSafeMigratableTables, ","), changes)
+		}
+	}
+
 	const maxConcurrentMigrate = 10
 	eg, ctx := errgroup.WithContext(ctx)
 	eg.SetLimit(maxConcurrentMigrate)
 
-	for _, table := range queries.NormalizedTables(tables) {
+	for _, table := range newSchema {
 		table := table
-		eg.Go(func() (err error) {
+		eg.Go(func() error {
 			c.logger.Info().Str("table", table.Name).Msg("Migrating table started")
-			defer func() {
-				c.logger.Err(err).Str("table", table.Name).Msg("Migrating table done")
-			}()
-
 			if len(table.Columns) == 0 {
 				c.logger.Warn().Str("table", table.Name).Msg("Table with no columns, skip")
 				return nil
 			}
 
-			curr := current.Get(table.Name)
-			if curr == nil {
+			current := currentSchema.Get(table.Name)
+			if current == nil {
 				return c.createTable(ctx, table)
 			}
 
-			return c.autoMigrateTable(ctx, curr, table)
+			err := c.autoMigrate(ctx, table, current)
+			if err != nil {
+				return err
+			}
+			c.logger.Err(err).Str("table", table.Name).Msg("Migrating table done")
```

Comment:
move back to defer

Reply:
Done in https://github.com/cloudquery/cloudquery/pull/8255/commits/e28b6ba3187e8139537290393d372f8e6a2f2e54